core/client/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -90,60 +231,165 @@ public void close() {
+      }
+    }
+
+    if (!mFileSystemWorkerClientHeartbeatPools.containsKey(rpcAddress)) {
+      FileSystemWorkerThriftClientPool pool =
+          new FileSystemWorkerThriftClientPool(mParentSubject, rpcAddress,
+              Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE_MAX),
+              Configuration.getLong(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_GC_THRESHOLD_MS));
+      if (mFileSystemWorkerClientHeartbeatPools.putIfAbsent(rpcAddress, pool) != null) {
+        pool.close();
+      }
+    }
+
     long sessionId = IdUtils.getRandomNonNegativeLong();
-    return new FileSystemWorkerClient(address, sessionId);
+    return new FileSystemWorkerClient(mFileSystemWorkerClientPools.get(rpcAddress),
+        mFileSystemWorkerClientHeartbeatPools.get(rpcAddress),
+        address, sessionId);
   }
 
   /**
-   * Releases a block master client into the block master client pool.
+   * Acquires a netty channel from the channel pools. If there is no available client instance
+   * available in the pool, it tries to create a new one. And an exception is thrown if it fails to
+   * create a new one.
    *
-   * @param masterClient a block master client to release
+   * @param address the network address of the channel
+   * @return the acquired netty channel
+   * @throws IOException if it fails to create a new client instance mostly because it fails to
+   *         connect to remote worker
    */
-  public void releaseMasterClient(FileSystemMasterClient masterClient) {
-    mFileSystemMasterClientPool.release(masterClient);
-  }
-
-  /**
-   * @return the Alluxio block store
-   */
-  public AlluxioBlockStore getAlluxioBlockStore() {
-    return mAlluxioBlockStore;
+  public Channel acquireNettyChannel(final InetSocketAddress address) throws IOException {
+    if (!mNettyChannelPools.containsKey(address)) {
+      Bootstrap bs = NettyClient.createClientBootstrap();
+      bs.remoteAddress(address);
+      NettyChannelPool pool = new NettyChannelPool(bs,
+          Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_CHANNEL_POOL_SIZE_MAX),
+          Configuration.getLong(PropertyKey.USER_NETWORK_NETTY_CHANNEL_POOL_GC_THRESHOLD_MS));
+      if (mNettyChannelPools.putIfAbsent(address, pool) != null) {
+        // This can happen if this function is called concurrently.
+        pool.close();
+      }
+    }
+    try {
+      return mNettyChannelPools.get(address).acquire();
+    } catch (InterruptedException e) {
+      throw Throwables.propagate(e);
+    }
   }
 
   /**
-   * @return the block store context
+   * Releases a netty channel to the channel pools.
+   *
+   * @param address the network address of the channel
+   * @param channel the channel to release
    */
-  public BlockStoreContext getBlockStoreContext() {
-    return mBlockStoreContext;
+  public void releaseNettyChannel(InetSocketAddress address, Channel channel) {
+    Preconditions.checkArgument(mNettyChannelPools.containsKey(address));
```

Comment:
If a channel is released after the pool is closed, would we hit this condition, and if so, should we handle it gracefully?